[iOS] Add build step to compile sandboxes

pvollan opened the pull request:
#### e5d7b0cb476dfb33eaf86d9452bff08722d528c0
<pre>
[iOS] Add build step to compile sandboxes
<a href="https://bugs.webkit.org/show_bug.cgi?id=262888">https://bugs.webkit.org/show_bug.cgi?id=262888</a>
rdar://116677504

Reviewed by Brent Fulgham and Elliott Williams.

Add build step to compile sandboxes on iOS to reduce the risk of committing sandbox changes that are not compiling.
This patch also fixes a few compile errors only seen with certain versions of the sandbox compiler tool.

* Source/WTF/wtf/PlatformHave.h:
* Source/WebKit/DerivedSources.make:
* Source/WebKit/GPUProcess/mac/com.apple.WebKit.GPUProcess.sb.in:
* Source/WebKit/NetworkProcess/mac/com.apple.WebKit.NetworkProcess.sb.in:
* Source/WebKit/WebProcess/com.apple.WebProcess.sb.in:

Canonical link: <a href="https://commits.webkit.org/269317@main">https://commits.webkit.org/269317@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/2d1e2f6e7770c670a62cd221d0c467efec730c35

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/22212 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/22416 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/23284 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/24109 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/20564 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/22460 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/26673 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/22740 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/21580 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/22445 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/22161 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/19274 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/24961 "Built successfully") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/39/builds/19218 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/20130 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/26378 "Passed tests") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/19315 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/20229 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/20358 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/24239 "Passed tests") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/21584 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/20793 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/17/builds/17695 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [❌ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/25630 "Failed to checkout and rebase branch from PR 18838") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/20145 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/19939 "Passed tests") | [❌ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/35/builds/25630 "Failed to checkout and rebase branch from PR 18838") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/24356 "Built successfully") | | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/24/builds/26904 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/2774 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/20741 "Built successfully") | | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/3/builds/5862 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

emw-apple commented on Source/WebKit/WebKit.xcodeproj/project.pbxproj:
```diff
@@ -17432,7 +17432,7 @@
 			runOnlyForDeploymentPostprocessing = 0;
 			shellPath = /bin/sh;
-			shellScript = "echo \"Preprocessing sandbox\"\nScripts/generate-derived-sources.sh sandbox-profiles-ios\nmkdir -p ${DSTROOT}/${INSTALL_PATH}\ncp ${BUILT_PRODUCTS_DIR}/DerivedSources/WebKit/com.apple.WebKit.adattributiond.sb ${DSTROOT}/${INSTALL_PATH}\ncp ${BUILT_PRODUCTS_DIR}/DerivedSources/WebKit/com.apple.WebKit.webpushd.sb ${DSTROOT}/${INSTALL_PATH}\ncp ${BUILT_PRODUCTS_DIR}/DerivedSources/ ...
+			shellScript = "echo \"Preprocessing sandbox\"\nScripts/generate-derived-sources.sh sandbox-profiles-ios\nmkdir -p ${DSTROOT}/${INSTALL_PATH}\n\nSANDBOX_GPU=${BUILT_PRODUCTS_DIR}/DerivedSources/WebKit/com.apple.WebKit.GPU.sb\nSANDBOX_NETWORKING=${BUILT_PRODUCTS_DIR}/DerivedSources/WebKit/com.apple.WebKit.Networking.sb\nSANDBOX_WEBCONTENT=${BUILT_PRODUCTS_DIR}/DerivedSources/WebKit/com.apple.Web ...
```
For anyone else looking at this, here's the shell script unescaped:
```
echo "Preprocessing sandbox"
Scripts/generate-derived-sources.sh sandbox-profiles-ios
mkdir -p ${DSTROOT}/${INSTALL_PATH}

SANDBOX_GPU=${BUILT_PRODUCTS_DIR}/DerivedSources/WebKit/com.apple.WebKit.GPU.sb
SANDBOX_NETWORKING=${BUILT_PRODUCTS_DIR}/DerivedSources/WebKit/com.apple.WebKit.Networking.sb
SANDBOX_WEBCONTENT=${BUILT_PRODUCTS_DIR}/DerivedSources/WebKit/com.apple.WebKit.WebContent.sb

sbutil compile ${SANDBOX_GPU}
sbutil compile ${SANDBOX_NETWORKING}
sbutil compile ${SANDBOX_WEBCONTENT}

cp ${SANDBOX_GPU} ${DSTROOT}/${INSTALL_PATH}
cp ${SANDBOX_NETWORKING} ${DSTROOT}/${INSTALL_PATH}
cp ${SANDBOX_WEBCONTENT} ${DSTROOT}/${INSTALL_PATH}

cp ${BUILT_PRODUCTS_DIR}/DerivedSources/WebKit/com.apple.WebKit.adattributiond.sb ${DSTROOT}/${INSTALL_PATH}
cp ${BUILT_PRODUCTS_DIR}/DerivedSources/WebKit/com.apple.WebKit.webpushd.sb ${DSTROOT}/${INSTALL_PATH}
cp ${BUILT_PRODUCTS_DIR}/DerivedSources/WebKit/com.apple.WebKit.WebAuthn.sb ${DSTROOT}/${INSTALL_PATH}
```

emw-apple replied:
We should probably be running this script with `set -e`, or manually `exit 1`-ing when the compile steps fail. 

My impression from reading the usage text is that `sbutil compile` outputs to stdout when not given an `-o` argument? We shouldn't spew a bunch of binary data into the log, so please >/dev/null at least :)

pvollan replied:
Those are good point, fixed in latest patch!

Thanks for reviewing!

brentfulgham replied:
I agree with both of Elliott's suggestions!

brentfulgham commented on Source/WebKit/WebKit.xcodeproj/project.pbxproj:
```diff
@@ -17432,7 +17431,7 @@
 			runOnlyForDeploymentPostprocessing = 0;
 			shellPath = /bin/sh;
-			shellScript = "echo \"Preprocessing sandbox\"\nScripts/generate-derived-sources.sh sandbox-profiles-ios\nmkdir -p ${DSTROOT}/${INSTALL_PATH}\ncp ${BUILT_PRODUCTS_DIR}/DerivedSources/WebKit/com.apple.WebKit.adattributiond.sb ${DSTROOT}/${INSTALL_PATH}\ncp ${BUILT_PRODUCTS_DIR}/DerivedSources/WebKit/com.apple.WebKit.webpushd.sb ${DSTROOT}/${INSTALL_PATH}\ncp ${BUILT_PRODUCTS_DIR}/DerivedSources/ ...
+			shellScript = "echo \"Preprocessing sandbox\"\n\nset -e\n\nScripts/generate-derived-sources.sh sandbox-profiles-ios\nmkdir -p ${DSTROOT}/${INSTALL_PATH}\n\nSANDBOX_GPU=${BUILT_PRODUCTS_DIR}/DerivedSources/WebKit/com.apple.WebKit.GPU.sb\nSANDBOX_NETWORKING=${BUILT_PRODUCTS_DIR}/DerivedSources/WebKit/com.apple.WebKit.Networking.sb\nSANDBOX_WEBCONTENT=${BUILT_PRODUCTS_DIR}/DerivedSources/WebKit/c ...
```
Do you need to tell the build script to include the `/usr/local/share/sandbox/profiles/embedded/imports/` path anywhere?

When I do the equivalent thing from the command line I have to have something like this in my shell environment:

```
IOS_SANDBOX_IMPORT_PATH="$DEV_ENVIRONMENT_PATH/Platforms/iPhoneOS.platform/Developer/SDKs/iPhoneOS$CURRENT_IOS_DEV_TARGET.Internal.sdk/usr/local/share/sandbox/profiles/embedded/imports/
```

brentfulgham reviewed (approved):
Looks great! Please consider whether the build environment knows about the Sandbox "import" path. Maybe it already does!

r=me

emw-apple commented on Source/WebKit/DerivedSources.make:
```diff
@@ -400,6 +400,9 @@ all : $(SANDBOX_PROFILES_WITHOUT_WEBPUSHD) $(WEBPUSHD_SANDBOX_PROFILE) $(SANDBOX
 	@echo Pre-processing $* sandbox profile...
 	grep -o '^[^;]*' $< | $(CC) $(SDK_FLAGS) $(TARGET_TRIPLE_FLAGS) $(SANDBOX_DEFINES) $(TEXT_PREPROCESSOR_FLAGS) $(FRAMEWORK_FLAGS) $(HEADER_FLAGS) $(EXTERNAL_FLAGS) -include "wtf/Platform.h" - > $@
+	if [ -e "/usr/local/bin/sbutil" ]; then \
+		sbutil compile -D HOMEDIR=homedir -D IMPORT_DIR=$(SDKROOT)/usr/local/share/sandbox/profiles/embedded/imports $@ \
```
Same as before, this shouldn't spew binary output into the build log.

Also, I'm assuming `HOMEDIR=homedir` is a fake directory, just needed to fill in a sandbox parameter?

emw-apple replied:
Oh, and the build error is because Make is escaping the newline on this line. So you need a `;` before the `fi`.

pvollan replied:
> Same as before, this shouldn't spew binary output into the build log.
> 
> Also, I'm assuming `HOMEDIR=homedir` is a fake directory, just needed to fill in a sandbox parameter?

Yes.

pvollan replied:
> Oh, and the build error is because Make is escaping the newline on this line. So you need a `;` before the `fi`.

Ah, thanks, will fix!

emw-apple reviewed (approved):
r+ with fix, thanks for this!

brentfulgham commented on Source/WebKit/GPUProcess/mac/com.apple.WebKit.GPUProcess.sb.in:
```diff
@@ -1057,6 +1057,19 @@
     (allow mach-kernel-endpoint
         (apply-message-filter
             (deny mach-message-send (with telemetry))
+#if PLATFORM(MAC) && __MAC_OS_X_VERSION_MIN_REQUIRED >= 140000
```
This would be better as a `HAVE` macro, rather than embed these specific version numbers in the file.

pvollan replied:
That is a good point. Will fix!

brentfulgham reviewed (approved):
Looks great! r=me

pvollan committed:
[iOS] Add build step to compile sandboxes
https://bugs.webkit.org/show_bug.cgi?id=262888
rdar://116677504

Reviewed by Brent Fulgham and Elliott Williams.

Add build step to compile sandboxes on iOS to reduce the risk of committing sandbox changes that are not compiling.
This patch also fixes a few compile errors only seen with certain versions of the sandbox compiler tool.

* Source/WTF/wtf/PlatformHave.h:
* Source/WebKit/DerivedSources.make:
* Source/WebKit/GPUProcess/mac/com.apple.WebKit.GPUProcess.sb.in:
* Source/WebKit/NetworkProcess/mac/com.apple.WebKit.NetworkProcess.sb.in:
* Source/WebKit/WebProcess/com.apple.WebProcess.sb.in:

Canonical link: https://commits.webkit.org/269317@main
